src/meta/src/barrier/backfill_order_control.rs

```diff
@@ -0,0 +1,149 @@
+    // The mapping between actors and fragment_ids
+    actor_to_fragment_id: HashMap<ActorId, FragmentId>,
+}
+
+// constructor
+impl BackfillOrderState {
+    pub fn new(
+        backfill_orders: HashMap<FragmentId, Vec<FragmentId>>,
+        stream_job_fragments: &StreamJobFragments,
+    ) -> Self {
+        tracing::debug!(?backfill_orders, "initialize backfill order state");
+        let actor_to_fragment_id = stream_job_fragments.actor_fragment_mapping();
+
+        let mut backfill_nodes: HashMap<FragmentId, BackfillNode> = HashMap::new();
+
+        for fragment in stream_job_fragments.fragments() {
+            if fragment.fragment_type_mask
+                & (FragmentTypeFlag::StreamScan as u32 | FragmentTypeFlag::SourceScan as u32)
+                > 0
+            {
+                let fragment_id = fragment.fragment_id;
+                backfill_nodes.insert(
+                    fragment_id,
+                    BackfillNode {
+                        fragment_id,
+                        remaining_actor_count: stream_job_fragments
+                            .fragment_actors(fragment_id)
+                            .len(),
+                        remaining_dependency_count: 0,
+                        children: vec![],
+                    },
+                );
+            }
+        }
+
+        for (fragment_id, children) in backfill_orders {
+            for child in &children {
+                let child_node = backfill_nodes.get_mut(child).unwrap();
+                child_node.remaining_dependency_count += 1;
+            }
+
+            let backfill_node = backfill_nodes.get_mut(&fragment_id).unwrap();
+            backfill_node.children = children;
+        }
+
+        let mut current_backfill_nodes = HashMap::new();
+        let mut remaining_backfill_nodes = HashMap::new();
+        for (fragment_id, node) in backfill_nodes {
+            if node.remaining_dependency_count == 0 {
+                current_backfill_nodes.insert(fragment_id, node);
+            } else {
+                remaining_backfill_nodes.insert(fragment_id, node);
+            }
+        }
+
+        Self {
+            current_backfill_nodes,
+            remaining_backfill_nodes,
+            actor_to_fragment_id,
+        }
+    }
+}
+
+// Getters
+impl BackfillOrderState {
+    pub fn get_initial_nodes(&self) -> Vec<FragmentId> {
+        self.current_backfill_nodes.keys().cloned().collect()
+    }
+}
+
+// state transitions
+impl BackfillOrderState {
+    pub fn finish_actor(&mut self, actor_id: ActorId) -> Vec<FragmentId> {
+        // Find the fragment_id of the actor.
+        let fragment_id = self.actor_to_fragment_id.get(&actor_id).unwrap();
+        // Decrease the remaining_actor_count of the operator.
+        // If the remaining_actor_count is 0, add the operator to the current_backfill_nodes.
+        let node = self.current_backfill_nodes.get_mut(fragment_id).unwrap();
```

Comment:
If that's the case, I think we can't unwrap the `current_backfill_nodes`, because source backfill could finish at any time.

Reply:
I have banned it in the binder first.

Reply:
https://github.com/risingwavelabs/risingwave/pull/20967/commits/217c501af6345c918d4a64270748a82e05decf6f